src/rigid.rs

```diff
@@ -0,0 +1,261 @@
+        //        = T2 * (R' * T' * R'^-1) * R'
+        //        = T2 * T'' * R'
+        //        = T'' * R'
+        //
+        // (R1^-1 * T1 * R1) = T' = T1 rotated by R1^-1
+        // R2 * R1  = R'
+        // (R' * T' * R'^-1) = T'' = T' rotated by R'
+        // T2 * T'' = T''' = T2 + T'
+
+        let t_prime = self
+            .rotation
+            .inverse()
+            .rotate_vector3d(&self.translation);
+        let r_prime = self.rotation.post_rotate(&other.rotation);
+        let t_prime2 = r_prime.rotate_vector3d(&t_prime);
+        let t_prime3 = t_prime2 + other.translation;
+        TypedRigidTransform3D {
+            rotation: r_prime,
+            translation: t_prime3,
+        }
+    }
+
+    /// Returns the multiplication of the two transforms such that
+    /// self's transformation applies after other's transformation.
+    ///
+    /// i.e., this produces `self * other`, premultiplying self to other
+    #[inline]
+    pub fn pre_mul<Src2>(
+        &self,
+        other: &TypedRigidTransform3D<T, Src2, Src>,
+    ) -> TypedRigidTransform3D<T, Src2, Dst> {
+        other.post_mul(&self)
+    }
+
+    /// Inverts the transformation
+    #[inline]
+    pub fn inverse(&self) -> TypedRigidTransform3D<T, Dst, Src> {
+        // result = (self)^-1
+        //        = (T * R)^-1
+        //        = R^-1 * T^-1
+        //        = R^-1 * T^-1 * R * R^-1
+        //        = (R^-1 * T^-1 * R) * R^-1
+        //        = T' * R'
+        //
+        // T' = (R^-1 * T^-1 * R) = (-T) rotated by R^-1
+        // R' = R^-1
+        //
+        // An easier way of writing this is to use new_from_reversed() with R^-1 and T^-1
+
+        TypedRigidTransform3D::new_from_reversed(
+            -self.translation,
+            self.rotation.inverse(),
+        )
+    }
+
+    pub fn to_transform(&self) -> TypedTransform3D<T, Src, Dst>
+    where
+        T: Trig,
+    {
+        self.translation
+            .to_transform()
+            .pre_mul(&self.rotation.to_transform())
+    }
+}
+
+impl<T: Float + ApproxEq<T>, Src, Dst> From<TypedRotation3D<T, Src, Dst>>
+    for TypedRigidTransform3D<T, Src, Dst>
+{
+    fn from(rot: TypedRotation3D<T, Src, Dst>) -> Self {
+        Self::from_rotation(rot)
+    }
+}
+
+impl<T: Float + ApproxEq<T>, Src, Dst> From<TypedVector3D<T, Dst>>
+    for TypedRigidTransform3D<T, Src, Dst>
+{
+    fn from(t: TypedVector3D<T, Dst>) -> Self {
+        Self::from_translation(t)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::RigidTransform3D;
+    use {Rotation3D, TypedTransform3D, Vector3D};
+
+    #[test]
+    fn test_rigid_construction() {
+        let translation = Vector3D::new(12.1, 17.8, -5.5);
+        let rotation = Rotation3D::unit_quaternion(0.5, -7.8, 2.2, 4.3);
+
+        let rigid = RigidTransform3D::new(rotation, translation);
+        assert!(rigid
+            .to_transform()
+            .approx_eq(&translation.to_transform().pre_mul(&rotation.to_transform())));
```

Comment:
Wait, it says in your comment above that `RigidTransform3D` is `Translation * Rotation`. But you have `Rotation * Translation` written here. `translation.pre_mul(rotation)` is `rotation.post_mul(translation)` and `A.post_mul(B)` is `AB`. See the definition of `post_mul` here: https://github.com/servo/euclid/blob/master/src/transform3d.rs#L259 (each entry is row of A times column of B).

Am I missing something or is this backwards?

Reply:
The type signatures of pre_mul and post_mul are backwards, then:

```rust
pub fn post_mul<NewDst>(&self, mat: &TypedTransform3D<T, Dst, NewDst>) -> TypedTransform3D<T, Src, NewDst>
```

i.e. `Transform<Src, Dst>.post_mul(Transform<Dst, NewDst>) = Transform<Src, Dst> * Transform<Dst, NewDst> = Transform<Src, NewDst>`. This doesn't make sense, if we were to transform a vector V the transformation would be `Transform<Src, Dst> * Transform<Dst, NewDst> * Vector<?>` , which forces the vector to be `Dst`, transforming it to `NewDst` and then somehow transforming it to `Dst` from `Src`, which it is not.

This all makes sense if the multiplication is backwards.

aside: `post_mul` and `pre_mul` are confusing method names because it's not clear if `a.post_mul(b)` means "a postmultiplied by b" or "postmultiply a with b" :smile:

Reply:
I think I got confused enough times with pre_mul and post_mul that I've verified what they do multiple times (we really should write the matrix math down in the docs) and I may have made a mistake with this, but also the typed units seem to work

Reply:
Just for laughs, Pathfinder and Euclid have `pre_mul` and `post_mul` swapped.

Reply:
So are the units wrong or are the names wrong? AB is each row of A times each column of B. That corresponds to `A.post_mul(B)` as written in the code fragment I linked, right? 

I think I'm going to abolish the `pre` and `post` stuff in Pathfinder. It's so confusing.

Reply:
Another possibility is that the docs are wrong and matrices are actually stored in column major order in Euclid. Sigh.

Reply:
> So are the units wrong or are the names wrong? A

The names are correct, the units are wrong.

Reply:
Wait, it's worse. Euclid is just doing math ... backwards.

```rust
    pub fn transform_vector3d(&self, v: &TypedVector3D<T, Src>) -> TypedVector3D<T, Dst> {
        vec3(
            v.x * self.m11 + v.y * self.m21 + v.z * self.m31,
            v.x * self.m12 + v.y * self.m22 + v.z * self.m32,
            v.x * self.m13 + v.y * self.m23 + v.z * self.m33,
        )
    }
```

"transform a vector [a b c] by a matrix T" is _usually_ `T * column matrix [a b c]`. However, this is `row matrix [a b c] * T`. Which seems broken?

Reply:
Well, not quite broken, it's totally valid to do transformations this way, but you need to change how you construct e.g. a rotation matrix.

Reply:
Yeah I think euclid doing matrix order wrong seems more correct.

Reply:
Typically I multiply matrices by column vectors in that order. i.e. Ma = b, where a and b are column vectors.

Reply:
Yes, that is what everyone does, but this is not what Euclid is doing, assuming it's using row-major order. Euclid is doing `aM=b` with row vectors.

Reply:
Also please don't take my word for anything I say about euclid's code here, I've been looking at this enough that it's quite possible that I'm missing something and just reading things wrong.

Reply:
Oh the _documentation_ of post_mul matches its units (and the way I use it, i.e. `a.post_mul(b) = ba`) but not the code:

```rust
/// Returns the multiplication of the two matrices such that mat's transformation
    /// applies after self's transformation.
    pub fn post_mul<NewDst>(&self, mat: &TypedTransform3D<T, Dst, NewDst>) -> TypedTransform3D<T, Src, NewDst> {
        TypedTransform3D::row_major(
```

"mat's transform applies after self's transform" means `mat * self`, which matches the units, and my interpretation. But not the code.

Increasingly convinced the code is column major

Reply:
```rust
    pub fn create_translation(x: T, y: T, z: T) -> Self {
        let (_0, _1): (T, T) = (Zero::zero(), One::one());
        TypedTransform3D::row_major(
            _1, _0, _0, _0,
            _0, _1, _0, _0,
            _0, _0, _1, _0,
             x,  y,  z, _1
        )
    }
```

oh hey the translation matrices are wrong too :)